File: leetcode/212-word-search-2.cpp
#include <cstdio>
#include <iostream>
#include <string>
#include <vector>
#include <set>

using namespace std;

class TrieNode {
public:
    int id;
    vector<int> next;
    TrieNode():id(-1){
        for (int i=0; i<26; i++) {
            next.push_back(-1);
        }
    }
    TrieNode(int id):id(id){
        for (int i=0; i<26; i++) {
            next.push_back(-1);
        }
    }
};

class TrieTree {
public:
    vector<TrieNode> pool;
    int root;

    void addWord(string s,int id) {
        int now = root;
        for(int i=0; i<s.length(); i++) {
            int x = s[i] - 'a';
            if (pool[now].next[x] < 0) {
                TrieNode tmp;
                pool.push_back(tmp);
                pool[now].next[x] = pool.size()-1;
            } 
            now = pool[now].next[x];

            if (i == s.length()-1) {
                pool[now].id = id;
            }
        }
    }

    void BuildTrie(vector<string> words) {
        TrieNode tmp;
        pool.push_back(tmp);
        root = 0;
        for(int i=0; i<words.size(); i++) {
            addWord(words[i],i);
        }
    }
};

class Solution {

public:
    TrieTree tree;
    set<int> ans;
    vector<vector<char> > dic;
    vector<vector<int> > vis;
    int dir[4][2] = {{1,0},{-1,0},{0,1},{0,-1}};
    int m,n;

    bool inboard(int x,int y) {
        return 0<=x && x<m && 0<=y && y<n;
    }

    void search(int x, int y, int node, int flag) {
        if (node < 0) {
            return;
        }
        if (tree.pool[node].id>=0) {
            ans.insert(tree.pool[node].id);
        }
        if (!inboard(x,y)) {
            return;
        }
        TrieNode now = tree.pool[node];
        if (now.id >= 0) {
        }
        for (int i=0; i<4; i++) {
            int xx = x+dir[i][0];
            int yy = y+dir[i][1];
            if (inboard(xx,yy) && vis[xx][yy] != flag) {
                vis[xx][yy] = flag;
                search(xx,yy,now.next[dic[xx][yy]-'a'],flag);
                vis[xx][yy] = -1;
            }
        }
    }

    vector<string> findWords(vector<vector<char> >& board, vector<string>& words) {
        vector<string> ans_v;
        if (board.size() <= 0) {
            return ans_v;
        }
        m = board.size();
        n = board[0].size();
        tree.BuildTrie(words);
        if (tree.pool.size() <= 0) {
            return ans_v;
        }
        dic = board;
        for (int i=0; i<board.size(); i++) {
            vector<int> tmp;
            for (int j=0; j<board[i].size(); j++) {
                tmp.push_back(-1);
            }
            vis.push_back(tmp);
        }
        int now = 0;
        for(int i=0; i<board.size(); i++) {
            for(int j=0; j<board[i].size(); j++) {
                vis[i][j] = now;
                search(i,j,tree.pool[0].next[dic[i][j]-'a'],now);
                now = now + 1;
            }
        }
        for (set<int>::iterator it=ans.begin(); it!=ans.end(); it++) {
            ans_v.push_back(words[*it]);
        }
        return ans_v;
    }
};

int main() {
    Solution s;
    vector<vector<char> > board;
    board.resize(2);
    board[0].push_back('a');
    board[0].push_back('b');
    board[1].push_back('d');
    board[1].push_back('c');
    vector<string> words;
    words.push_back("abcd");
    vector<string> ans = s.findWords(board,words);
    for(int i=0; i<ans.size(); i++) {
        cout<<ans[i]<<endl;
    }
    return 0;
}

File: leetcode/212-word-search-2.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

class TreeNode(object):
    is_tail = False
    idx = -1 
    childrens = {}

class TrieTree(object):
    head = TreeNode()

    def add_word(self, word, word_idx):
        cur_node = self.head
        for idx, c in enumerate(word):
            if c not in cur_node.childrens:
                cur_node.childrens[c] = TreeNode()
                cur_node.childrens[c].childrens = {}
            if idx == len(word) - 1:
                cur_node.childrens[c].idx = word_idx
                cur_node.childrens[c].is_tail = True
            cur_node = cur_node.childrens[c]

class Solution:
    bm = 0
    bn = 0
    board = None
    vis = []
    vis_idx = 0
    find_words = None

    directions = [(0, 1), (0, -1), (-1, 0), (1, 0)]

    def in_board(self, x, y):
        if x < 0 or x >= self.bm:
            return False
        if y < 0 or y >= self.bn:
            return False
        return True

    def dfs(self, x, y, node):
        if node.is_tail:
            self.find_words.add(node.idx)
        #print(node.childrens)
        #print(x, y, self.board[x][y], node.is_tail, node.idx, self.vis_idx)
        self.vis[x][y] = self.vis_idx

        for dx, dy in self.directions:
            nx = dx + x
            ny = dy + y
            if not self.in_board(nx, ny):
                continue
            if self.vis[nx][ny] == self.vis_idx:
                continue
            if self.board[nx][ny] not in node.childrens:
                continue
            next_node = node.childrens[self.board[nx][ny]]
            self.dfs(nx, ny, next_node)
        self.vis[x][y] = 0

    def findWords(self, board, words):
        """
        :type board: List[List[str]]
        :type words: List[str]
        :rtype: List[str]
        """
        self.find_words = set()
        self.bm = len(board)
        self.bn = len(board[0])
        self.board = board
        if self.bm == 0:
            return []
        self.vis = [[0 for x in range(self.bn)] \
            for y in range(self.bm)]
        tt = TrieTree()
        tt.head.childrens = {}
        for idx, word in enumerate(words):
            tt.add_word(word, idx)
        self.vis_idx = 1
        for i in range(self.bm):
            for j in range(self.bn):
                cur = self.board[i][j]
                if cur in tt.head.childrens:
                    self.dfs(i, j, tt.head.childrens[cur])
                    self.vis_idx = self.vis_idx + 1
        find_words_str = []
        for idx in self.find_words:
            if idx < 0 or idx >= len(words):
                continue
            find_words_str.append(words[idx])
        return find_words_str

board = [
  ['o','a','a','n'],
  ['e','t','a','e'],
  ['i','h','k','r'],
  ['i','f','l','v']
]
solution = Solution()
"""
finds = solution.findWords(board, ["oath","pea","eat","rain"])
print(finds)
finds = solution.findWords([['a']], ['a'])
print(finds)
finds = solution.findWords([['a']], ['ab'])
print(finds)
board = [["a","b","c"],["a","e","d"],["a","f","g"]]
arr = ["abcdefg","gfedcbaaa","eaabcdgfa","befa","dgc","ade"]
finds = solution.findWords(board, arr)
print(finds)
"""
board = [["a","b"],["a","a"]]
arr = ["aba","baa","bab","aaab","aaa","aaaa","aaba"]
finds = solution.findWords(board, arr)
print(finds)
